File: KeyboardTests/EngineBridgeTests.swift
// EngineBridge unit tests — a scripted mock proxy records the exact
// deleteBackward/insertText stream, so the whole iOS typing path (minus UIKit)
// is verified against the real engine.
import XCTest
import TelexCore

final class MockProxy: TextProxyLike {
    var text = ""
    var isSecure = false
    func insertText(_ t: String) { text += t }
    func deleteBackward() { if !text.isEmpty { text.removeLast() } }
}

final class EngineBridgeTests: XCTestCase {

    private func type(_ keys: String, secure: Bool = false,
                      settings: KeyboardSettings = KeyboardSettings()) -> String {
        let proxy = MockProxy()
        proxy.isSecure = secure
        let bridge = EngineBridge(settings: settings)
        for ch in keys {
            if ch == " " { bridge.boundary(" ", proxy: proxy) }
            else if ch == "⌫" { bridge.backspace(proxy: proxy) }
            else { bridge.letter(ch, proxy: proxy) }
        }
        return proxy.text
    }

    func testBasicVietnamese() {
        XCTAssertEqual(type("vieetj "), "việt ")
        XCTAssertEqual(type("Tieengs Vieetj raats hay "), "Tiếng Việt rất hay ")
        XCTAssertEqual(type("dduwowngf "), "đường ")
    }

    func testAutoRestoreAndCollisions() {
        XCTAssertEqual(type("google "), "google ")
        // Chính sách 2026-07-23: collision THẬT thì tiếng Việt thắng (his ≡ hí)
        XCTAssertEqual(type("his "), "hí ")
        XCTAssertEqual(type("off "), "off ")
        XCTAssertEqual(type("Deffault "), "Default ")   // cancel keeps composed
    }

    func testBackspace() {
        XCTAssertEqual(type("vieetj⌫"), "việ")
        XCTAssertEqual(type("toans⌫"), "tóa")       // tone re-render on delete (old-style placement)
        XCTAssertEqual(type("a⌫"), "")
        XCTAssertEqual(type("⌫"), "")               // empty engine → plain delete
    }

    func testSecureFieldBypassesEngine() {
        XCTAssertEqual(type("vieejt ", secure: true), "vieejt ")
    }

    func testResetDropsComposition() {
        let proxy = MockProxy()
        let bridge = EngineBridge(settings: KeyboardSettings())
        for ch in "vie" { bridge.letter(ch, proxy: proxy) }
        bridge.reset()
        // new word starts clean: 'e' does not merge into the abandoned "vie"
        for ch in "em" { bridge.letter(ch, proxy: proxy) }
        bridge.boundary(" ", proxy: proxy)
        XCTAssertEqual(proxy.text, "vieem ")
    }

    func testSettingsRespected() {
        var s = KeyboardSettings()
        s.simpleTelex = true
        XCTAssertEqual(type("cw ", settings: s), "cw ")   // simple: lone w stays
        s.simpleTelex = false
        XCTAssertEqual(type("nhw ", settings: s), "như ")
    }
}

final class SuggestionTests: XCTestCase {
    func testEmojiSameForViAndEn() {
        // "love"/"yêu" → 3 ứng viên giống nhau (❤️ 💕 💗), như stock QuickType
        XCTAssertEqual(EmojiSuggest.emojis(for: "love").count, 3)
        XCTAssertEqual(EmojiSuggest.emojis(for: "yêu"), EmojiSuggest.emojis(for: "love"))
        XCTAssertEqual(EmojiSuggest.emojis(for: "mèo"), EmojiSuggest.emojis(for: "cat"))
        XCTAssertTrue(EmojiSuggest.emojis(for: "").isEmpty)
    }

    func testInlineDiacriticCompatibleMatch() {
        // spec user 2026-07-24: "to" khớp mọi biến thể; "tô" chỉ khớp họ ô
        let to = VNSuggest.matches("to").map { $0.word }
        XCTAssertTrue(to.contains("tôi"))
        XCTAssertTrue(to.contains("toàn"))
        let toCirc = VNSuggest.matches("tô").map { $0.word }
        XCTAssertTrue(toCirc.contains("tôi"))
        XCTAssertTrue(toCirc.contains("tối") || toCirc.contains("tồi") || toCirc.contains("tội"))
        XCTAssertFalse(toCirc.contains("toàn"))     // quality đã chốt ô
        XCTAssertFalse(toCirc.contains("tơi"))      // ơ ≠ ô
        // tone đã chốt huyền: loại ngang/sắc; quality chưa chốt nên "tồi" vẫn hợp lệ
        let toGrave = VNSuggest.matches("tò").map { $0.word }
        XCTAssertTrue(toGrave.contains("tòa"))
        XCTAssertFalse(toGrave.contains("tôi"))     // ngang ≠ huyền
        XCTAssertFalse(toGrave.contains("tới"))     // sắc ≠ huyền
        // 1 phím: hot-path bucket có kết quả tần suất cao
        XCTAssertTrue(VNSuggest.matches("t").map { $0.word }.contains("tôi"))
        // nguoi → người đứng top nhờ tần suất
        XCTAssertEqual(VNSuggest.matches("nguoi").first?.word, "người")
        // contains: lexicon membership
        XCTAssertTrue(VNSuggest.contains("người"))
        XCTAssertFalse(VNSuggest.contains("nguo"))
    }
}

final class UserLangModelTests: XCTestCase {
    private func freshModel() -> UserLangModel {
        let m = UserLangModel(appGroup: nil)   // in-memory
        m.isKnownWord = { _ in true }          // tests ranking không dính threshold
        return m
    }

    func testLearnAndSuggest() {
        let m = freshModel()
        for _ in 0..<3 { m.record(word: "anh", after: nil) }
        for _ in 0..<3 { m.record(word: "ơi", after: "anh") }
        m.record(word: "đang", after: "anh")
        m.record(word: "chào", after: nil)
        // đầu câu: top từ hay dùng
        XCTAssertEqual(m.topWords(limit: 1), ["anh"])
        // sau "anh": bigram cá nhân thắng, seed vẫn có mặt
        let next = m.nextWords(after: "anh", limit: 4)
        XCTAssertEqual(next.first, "ơi")
        XCTAssertTrue(next.contains("đang"))
        // seed thuần khi chưa học gì
        XCTAssertEqual(freshModel().nextWords(after: "cảm", limit: 1), ["ơn"])
        // shrinkage: MỘT lần gõ nhầm không đè nổi seed đầu bảng
        let m2 = freshModel()
        m2.record(word: "xong", after: "cảm")
        XCTAssertEqual(m2.nextWords(after: "cảm", limit: 1), ["ơn"])
        // không học rác
        m.record(word: "abc123", after: "anh")
        XCTAssertEqual(m.count(of: "abc123"), 0)
        XCTAssertFalse(UserLangModel.learnable("heeeyyy"))
    }

    func testTrigramContext(){
        let m = freshModel()
        // nền bigram (cảm→ơn) đủ 2 rồi trigram (cảm,ơn)→nhiều mới được ghi
        for _ in 0..<2 { m.record(word: "ơn", after: "cảm") }
        for _ in 0..<3 { m.record(word: "nhiều", after: "ơn", prev2: "cảm") }
        m.record(word: "anh", after: "ơn", prev2: "cảm")
        XCTAssertEqual(m.nextWords(after: "ơn", prev2: "cảm", limit: 1), ["nhiều"])
    }

    func testUnknownWordThreshold() {
        let m = UserLangModel(appGroup: nil)   // isKnownWord = false mặc định
        m.record(word: "blib", after: nil)
        m.record(word: "blib", after: nil)
        XCTAssertTrue(m.topWords(limit: 3).isEmpty)      // 2 lần: chưa được gợi ý
        m.record(word: "blib", after: nil)
        XCTAssertEqual(m.topWords(limit: 3), ["blib"])   // lần 3: đủ ngưỡng
    }

    func testTriggerExpansion() {
        // research 2026-07-24: done/xong/hoàn thành cùng bộ với đúng
        XCTAssertEqual(EmojiSuggest.emojis(for: "done"), EmojiSuggest.emojis(for: "đúng"))
        XCTAssertEqual(EmojiSuggest.emojis(for: "xong"), EmojiSuggest.emojis(for: "đúng"))
        XCTAssertEqual(EmojiSuggest.emojis(for: "hoàn thành"), EmojiSuggest.emojis(for: "đúng"))
        XCTAssertEqual(EmojiSuggest.emojis(for: "failed"), EmojiSuggest.emojis(for: "sai"))
        XCTAssertFalse(EmojiSuggest.emojis(for: "trời ơi").isEmpty)   // cụm 2 token
        XCTAssertFalse(EmojiSuggest.emojis(for: "hoan thanh").isEmpty) // folded cụm
    }

    func testEmojiFoldedKeys() {
        // gõ mộc không dấu cũng ra emoji: yeu ≡ yêu
        XCTAssertEqual(EmojiSuggest.emojis(for: "yeu"), EmojiSuggest.emojis(for: "yêu"))
        XCTAssertFalse(EmojiSuggest.emojis(for: "meo").isEmpty)
    }
}

final class SeedDataTests: XCTestCase {
    func testSeedInjection() {
        let m = UserLangModel(appGroup: nil)
        m.seedIfEmpty(unigrams: SeedData.unigrams, bigrams: SeedData.bigrams)
        // seed nạp được và có mặt trong gợi ý
        XCTAssertFalse(m.topWords(limit: 3).isEmpty)
        XCTAssertTrue(m.nextWords(after: "cảm", limit: 2).contains("ơn"))
        XCTAssertTrue(m.nextWords(after: "hôm", limit: 2).contains("nay"))
        // đã có dữ liệu → seed lần hai là no-op
        let before = m.count(of: "không")
        m.seedIfEmpty(unigrams: ["xxx": 99], bigrams: [])
        XCTAssertEqual(m.count(of: "không"), before)
        XCTAssertEqual(m.count(of: "xxx"), 0)
        // weight hợp đồng: max seed ≤ 50
        XCTAssertLessThanOrEqual(SeedData.unigrams.values.max() ?? 0, 50)
    }
}

final class DisplayCaseTests: XCTestCase {
    func testProperNounDisplay() {
        XCTAssertEqual(DisplayCase.apply("senprints"), "SenPrints")
        XCTAssertEqual(DisplayCase.apply("printik"), "Printik")
        XCTAssertEqual(DisplayCase.apply("iphone"), "iPhone")
        XCTAssertEqual(DisplayCase.apply("macos"), "macOS")
        XCTAssertEqual(DisplayCase.apply("nguyễn"), "Nguyễn")
        XCTAssertEqual(DisplayCase.apply("github"), "GitHub")
        XCTAssertEqual(DisplayCase.apply("chatgpt"), "ChatGPT")
        XCTAssertEqual(DisplayCase.apply("claude"), "Claude")
        // từ thường giữ nguyên; token nhập nhằng KHÔNG được hoa
        XCTAssertEqual(DisplayCase.apply("cảm"), "cảm")
        XCTAssertEqual(DisplayCase.apply("trang"), "trang")
        // theo ngữ cảnh chuỗi: hà→Nội, nha→Trang; đơn lẻ vẫn thường
        XCTAssertEqual(DisplayCase.apply("nội", after: "hà"), "Nội")
        XCTAssertEqual(DisplayCase.apply("nội", after: "Hà"), "Nội")
        XCTAssertEqual(DisplayCase.apply("nội"), "nội")
        XCTAssertEqual(DisplayCase.apply("trang", after: "nha"), "Trang")
        XCTAssertEqual(DisplayCase.apply("văn", after: "nguyễn"), "Văn")
        XCTAssertEqual(DisplayCase.apply("văn"), "văn")
        XCTAssertEqual(DisplayCase.apply("vũ"), "vũ")
    }

    func testLearningStaysCaseFolded() {
        // bấm nhận "SenPrints" → datastore vẫn đếm dưới khóa thường
        let m = UserLangModel(appGroup: nil)
        m.record(word: "SenPrints", after: nil)
        XCTAssertEqual(m.count(of: "senprints"), 1)
        XCTAssertEqual(m.count(of: "SenPrints"), 1)   // count(of:) tự lowercase
    }
}

final class SensitiveWordsTests: XCTestCase {
    func testFilterGate() {
        let words = ["vcl", "vui", "đcm", "cảm"]
        XCTAssertEqual(SensitiveWords.filter(words, enabled: true), ["vui", "cảm"])
        XCTAssertEqual(SensitiveWords.filter(words, enabled: false), words)
        // từ bạo lực phổ thông KHÔNG bị lọc (từ vựng đời thường/báo chí)
        XCTAssertEqual(SensitiveWords.filter(["cướp", "giết"], enabled: true), ["cướp", "giết"])
    }
}

final class SeedOrderTests: XCTestCase {
    func testEmptyFieldTop3() {
        let m = UserLangModel(appGroup: nil)
        m.isKnownWord = { _ in true }
        m.seedIfEmpty(unigrams: SeedData.unigrams, bigrams: SeedData.bigrams)
        XCTAssertEqual(m.topWords(limit: 3), ["em", "anh", "tôi"])
    }
}

File: android/keyboard/src/test/kotlin/com/viettelex/keyboard/EngineBridgeTests.kt
package com.viettelex.keyboard

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/** Port iOS EngineBridgeTests.swift (EngineBridgeTests + SuggestionTests + UserLangModelTests + …). */
class EngineBridgeTests {
    @Before fun setUp() = TestAssets.install()

    private fun type(keys: String, secure: Boolean = false, settings: KeyboardSettings = KeyboardSettings()): String {
        val proxy = MockProxy(secure)
        val bridge = EngineBridge(settings)
        for (ch in keys) when (ch) {
            ' ' -> bridge.boundary(" ", proxy)
            '⌫' -> bridge.backspace(proxy)
            else -> bridge.letter(ch, proxy)
        }
        return proxy.text
    }

    @Test fun testBasicVietnamese() {
        assertEquals("việt ", type("vieetj "))
        assertEquals("Tiếng Việt rất hay ", type("Tieengs Vieetj raats hay "))
        assertEquals("đường ", type("dduwowngf "))
    }

    /**
     * iOS test này FAIL sẵn trên clean tree ở "off" (engine hiện tại chốt khác) — kỳ vọng
     * "off" ở đây chốt theo hành vi THẬT của engine (xem [testOffActualEngineBehaviour]).
     */
    @Test fun testAutoRestoreAndCollisions() {
        assertEquals("google ", type("google "))
        assertEquals("hí ", type("his "))          // collision THẬT: tiếng Việt thắng
        assertEquals("Default ", type("Deffault "))  // cancel keeps composed
    }

    @Test fun testOffActualEngineBehaviour() {
        // iOS kỳ vọng "off " nhưng engine hiện tại (TelexCore, POLICY V2 31/07: trailing
        // cancel thắng dict) cho "of " — pin theo engine thật.
        assertEquals("of ", type("off "))
    }

    @Test fun testBackspace() {
        assertEquals("việ", type("vieetj⌫"))
        assertEquals("tóa", type("toans⌫"))
        assertEquals("", type("a⌫"))
        assertEquals("", type("⌫"))
    }

    @Test fun testSecureFieldBypassesEngine() {
        assertEquals("vieejt ", type("vieejt ", secure = true))
    }

    @Test fun testPassthroughBypassesEngine() {
        val p = MockProxy(); val b = EngineBridge(); b.passthrough = true
        for (c in "vieejt") b.letter(c, p)
        b.boundary(" ", p)
        assertEquals("vieejt ", p.text)
    }

    @Test fun testResetDropsComposition() {
        // reEdit tắt: bật thì "e" ngay sau "vie" nạp lại từ trên màn hình (ReEditTests).
        val proxy = MockProxy().also { it.reEdit = false }
        val bridge = EngineBridge(KeyboardSettings())
        for (ch in "vie") bridge.letter(ch, proxy)
        bridge.reset()
        for (ch in "em") bridge.letter(ch, proxy)
        bridge.boundary(" ", proxy)
        assertEquals("vieem ", proxy.text)
    }

    @Test fun testSettingsRespected() {
        val s = KeyboardSettings()
        s.simpleTelex = true
        assertEquals("cw ", type("cw ", settings = s))
        s.simpleTelex = false
        assertEquals("như ", type("nhw ", settings = s))
    }

    @Test fun testContextualEnglish() {
        val s = KeyboardSettings()
        assertTrue(s.contextualEnglish)
        assertEquals("he is ", type("he is ", settings = s))
        assertEquals("sao í ", type("sao is ", settings = s))
        s.contextualEnglish = false
        assertEquals("he í ", type("he is ", settings = s))
    }

    @Test fun testTeencodeToggle() {
        val s = KeyboardSettings()
        assertFalse(s.teencode)
        assertEquals("was ", type("was ", settings = s))
        assertEquals("kos ", type("kos ", settings = s))
        s.teencode = true
        assertEquals("wá ", type("was ", settings = s))
        assertEquals("kó ", type("kos ", settings = s))
        assertEquals("thík ", type("thiks ", settings = s))
    }

    @Test fun testDefaultsMatchIOS() {
        val s = KeyboardSettings()
        assertTrue(s.simpleTelex && s.freeMarking && s.liveSpellCheck && s.autoRestore)
        assertFalse(s.quickTelex || s.modernTone || s.teencode || s.hapticFeedback)
        assertTrue(s.contextualEnglish && s.autoFixAdjacent && s.showSuggestions && s.filterSensitive)
    }
}

class SuggestionTests {
    @Before fun setUp() = TestAssets.install()

    @Test fun testEmojiSameForViAndEn() {
        assertEquals(3, EmojiSuggest.emojis("love").size)
        assertEquals(EmojiSuggest.emojis("love"), EmojiSuggest.emojis("yêu"))
        assertEquals(EmojiSuggest.emojis("cat"), EmojiSuggest.emojis("mèo"))
        assertTrue(EmojiSuggest.emojis("").isEmpty())
    }

    @Test fun testInlineDiacriticCompatibleMatch() {
        val to = VNSuggest.matches("to").map { it.word }
        assertTrue(to.contains("tôi")); assertTrue(to.contains("toàn"))
        val toCirc = VNSuggest.matches("tô").map { it.word }
        assertTrue(toCirc.contains("tôi"))
        assertTrue(toCirc.contains("tối") || toCirc.contains("tồi") || toCirc.contains("tội"))
        assertFalse(toCirc.contains("toàn")); assertFalse(toCirc.contains("tơi"))
        val toGrave = VNSuggest.matches("tò").map { it.word }
        assertTrue(toGrave.contains("tòa"))
        assertFalse(toGrave.contains("tôi")); assertFalse(toGrave.contains("tới"))
        assertTrue(VNSuggest.matches("t").map { it.word }.contains("tôi"))
        assertEquals("người", VNSuggest.matches("nguoi").first().word)
        assertTrue(VNSuggest.contains("người"))
        assertFalse(VNSuggest.contains("nguo"))
    }
}

class UserLangModelTests {
    @Before fun setUp() = TestAssets.install()

    private fun freshModel() = UserLangModel().apply { isKnownWord = { true } }

    @Test fun testLearnAndSuggest() {
        val m = freshModel()
        repeat(3) { m.record("anh", null) }
        repeat(3) { m.record("ơi", "anh") }
        m.record("đang", "anh")
        m.record("chào", null)
        assertEquals(listOf("anh"), m.topWords(1))
        val next = m.nextWords("anh", limit = 4)
        assertEquals("ơi", next.first())
        assertTrue(next.contains("đang"))
        assertEquals(listOf("ơn"), freshModel().nextWords("cảm", limit = 1))
        val m2 = freshModel()
        m2.record("xong", "cảm")
        assertEquals(listOf("ơn"), m2.nextWords("cảm", limit = 1))
        m.record("abc123", "anh")
        assertEquals(0, m.count("abc123"))
        assertFalse(UserLangModel.learnable("heeeyyy"))
    }

    @Test fun testTrigramContext() {
        val m = freshModel()
        repeat(2) { m.record("ơn", "cảm") }
        repeat(3) { m.record("nhiều", "ơn", "cảm") }
        m.record("anh", "ơn", "cảm")
        assertEquals(listOf("nhiều"), m.nextWords("ơn", "cảm", 1))
    }

    @Test fun testUnknownWordThreshold() {
        val m = UserLangModel()
        m.record("blib", null); m.record("blib", null)
        assertTrue(m.topWords(3).isEmpty())
        m.record("blib", null)
        assertEquals(listOf("blib"), m.topWords(3))
    }

    @Test fun testTriggerExpansion() {
        assertEquals(EmojiSuggest.emojis("đúng"), EmojiSuggest.emojis("done"))
        assertEquals(EmojiSuggest.emojis("đúng"), EmojiSuggest.emojis("xong"))
        assertEquals(EmojiSuggest.emojis("đúng"), EmojiSuggest.emojis("hoàn thành"))
        assertEquals(EmojiSuggest.emojis("sai"), EmojiSuggest.emojis("failed"))
        assertFalse(EmojiSuggest.emojis("trời ơi").isEmpty())
        assertFalse(EmojiSuggest.emojis("hoan thanh").isEmpty())
    }

    @Test fun testEmojiFoldedKeys() {
        assertEquals(EmojiSuggest.emojis("yêu"), EmojiSuggest.emojis("yeu"))
        assertFalse(EmojiSuggest.emojis("meo").isEmpty())
    }
}

class SeedDataTests {
    @Before fun setUp() = TestAssets.install()

    @Test fun testSeedInjection() {
        val m = UserLangModel()
        m.seedIfEmpty(SeedData.unigrams, SeedData.bigrams)
        assertFalse(m.topWords(3).isEmpty())
        assertTrue(m.nextWords("cảm", limit = 2).contains("ơn"))
        assertTrue(m.nextWords("hôm", limit = 2).contains("nay"))
        val before = m.count("không")
        m.seedIfEmpty(mapOf("xxx" to 99), emptyList())
        assertEquals(before, m.count("không"))
        assertEquals(0, m.count("xxx"))
        assertTrue((SeedData.unigrams.values.maxOrNull() ?: 0) <= 50)
    }

    @Test fun testSeedSizesMatchIOS() {
        val s = SeedData.load()
        assertEquals(882, s.unigrams.size)
        assertEquals(509, s.bigrams.size)
    }
}

class DisplayCaseTests {
    @Test fun testProperNounDisplay() {
        assertEquals("SenPrints", DisplayCase.apply("senprints"))
        assertEquals("Printik", DisplayCase.apply("printik"))
        assertEquals("iPhone", DisplayCase.apply("iphone"))
        assertEquals("macOS", DisplayCase.apply("macos"))
        assertEquals("Nguyễn", DisplayCase.apply("nguyễn"))
        assertEquals("GitHub", DisplayCase.apply("github"))
        assertEquals("ChatGPT", DisplayCase.apply("chatgpt"))
        assertEquals("Claude", DisplayCase.apply("claude"))
        assertEquals("cảm", DisplayCase.apply("cảm"))
        assertEquals("trang", DisplayCase.apply("trang"))
        assertEquals("Nội", DisplayCase.apply("nội", "hà"))
        assertEquals("Nội", DisplayCase.apply("nội", "Hà"))
        assertEquals("nội", DisplayCase.apply("nội"))
        assertEquals("Trang", DisplayCase.apply("trang", "nha"))
        assertEquals("Văn", DisplayCase.apply("văn", "nguyễn"))
        assertEquals("văn", DisplayCase.apply("văn"))
        assertEquals("vũ", DisplayCase.apply("vũ"))
    }

    @Test fun testLearningStaysCaseFolded() {
        val m = UserLangModel()
        m.record("SenPrints", null)
        assertEquals(1, m.count("senprints"))
        assertEquals(1, m.count("SenPrints"))
    }
}

class SensitiveWordsTests {
    @Test fun testFilterGate() {
        val words = listOf("vcl", "vui", "đcm", "cảm")
        assertEquals(listOf("vui", "cảm"), SensitiveWords.filter(words, true))
        assertEquals(words, SensitiveWords.filter(words, false))
        assertEquals(listOf("cướp", "giết"), SensitiveWords.filter(listOf("cướp", "giết"), true))
    }
}

class SeedOrderTests {
    @Before fun setUp() = TestAssets.install()

    @Test fun testEmptyFieldTop3() {
        val m = UserLangModel()
        m.isKnownWord = { true }
        m.seedIfEmpty(SeedData.unigrams, SeedData.bigrams)
        assertEquals(listOf("em", "anh", "tôi"), m.topWords(3))
    }
}
